lib/filesystem.nix

```diff
@@ -47,11 +47,13 @@
   # listFilesRecursive: Path -> [ Path ]
   #
   # Given a directory, return a flattened list of all files within it recursively.
-  listFilesRecursive = dir: lib.flatten (lib.mapAttrsToList (name: type:
-    if type == "directory" then
-      lib.filesystem.listFilesRecursive (dir + "/${name}")
-    else
-      dir + "/${name}"
-  ) (builtins.readDir dir));
+  listFilesRecursive = dir: let
+    recurse = dir: lib.flatten (lib.mapAttrsToList (name: type:
```

Comment:
This `dir` shadows the outer `dir`, which isn't great.
The outer `dir` isn't even needed though, at which point you might as well remove the let binding, and we're back to square one.
To me, this implementation is more complicated and the steps required to reach this form are simple refactoring techniques. I don't think it's the purpose of this function's implementation to be something other than the simplest definition it can reasonably be.